File: examples/cpp/masked_operator.cpp
#include <cstddef>
#include <cstdint>
#include <vector>

#include <tsl.hpp>

struct less_than_op {
    template <class Vec>
    typename Vec::mask_type operator()(
        typename tsl::reg_param<Vec>::type left,
        typename tsl::reg_param<Vec>::type right) const {
        return tsl::less_than<Vec>(left, right);
    }
};

struct square_or_original_op {
    template <class Vec>
    typename Vec::register_type operator()(
        typename Vec::mask_type active,
        typename tsl::reg_param<Vec>::type value) const {
        const auto squared = tsl::mul<Vec>(value, value);
        return tsl::blend<Vec>(active, value, squared);
    }
};

struct add_or_left_op {
    template <class Vec>
    typename Vec::register_type operator()(
        typename Vec::mask_type active,
        typename tsl::reg_param<Vec>::type left,
        typename tsl::reg_param<Vec>::type right) const {
        const auto sum = tsl::add<Vec>(left, right);
        return tsl::blend<Vec>(active, left, sum);
    }
};

void fill_inputs(
    std::vector<std::int32_t>& left,
    std::vector<std::int32_t>& right) {
    for (std::size_t i = 0; i < left.size(); ++i) {
        left[i] = static_cast<std::int32_t>(static_cast<int>(i % 29) - 14);
        right[i] = static_cast<std::int32_t>(static_cast<int>((i * 11) % 47) - 23);
    }
}

template <class Parallelism>
bool run_masked_unary_case() {
    constexpr std::size_t count = 1000;
    constexpr std::int32_t sentinel = -777777;
    std::vector<std::int32_t> input(count);
    std::vector<std::int32_t> threshold(count);
    std::vector<std::int32_t> output(count, sentinel);
    fill_inputs(input, threshold);

    using mask_type = tsl::algo::integral_mask_type<Parallelism, std::int32_t>;
    std::vector<mask_type> masks(
        tsl::algo::integral_mask_chunk_count<Parallelism, std::int32_t>(count));

    tsl::algo::predicate_binary<Parallelism, tsl::algo::alignment::unaligned>(
        less_than_op{},
        input.data(),
        threshold.data(),
        masks.data(),
        count);
    tsl::algo::transform_masked_unary<Parallelism, tsl::algo::alignment::unaligned>(
        square_or_original_op{},
        input.data(),
        masks.data(),
        output.data(),
        count);

    for (std::size_t i = 0; i < count; ++i) {
        const bool active = input[i] < threshold[i];
        const auto expected = active ? input[i] * input[i] : input[i];
        if (output[i] != expected) {
            return false;
        }
    }
    return true;
}

template <class Parallelism>
bool run_masked_binary_case() {
    constexpr std::size_t count = 1000;
    constexpr std::int32_t sentinel = -888888;
    std::vector<std::int32_t> left(count);
    std::vector<std::int32_t> right(count);
    std::vector<std::int32_t> output(count, sentinel);
    fill_inputs(left, right);

    using mask_type = tsl::algo::integral_mask_type<Parallelism, std::int32_t>;
    std::vector<mask_type> masks(
        tsl::algo::integral_mask_chunk_count<Parallelism, std::int32_t>(count));

    tsl::algo::predicate_binary<Parallelism, tsl::algo::alignment::unaligned>(
        less_than_op{},
        left.data(),
        right.data(),
        masks.data(),
        count);
    tsl::algo::transform_masked_binary<Parallelism, tsl::algo::alignment::unaligned>(
        add_or_left_op{},
        left.data(),
        right.data(),
        masks.data(),
        output.data(),
        count);

    for (std::size_t i = 0; i < count; ++i) {
        const bool active = left[i] < right[i];
        const auto expected = active ? left[i] + right[i] : left[i];
        if (output[i] != expected) {
            return false;
        }
    }
    return true;
}

template <class Parallelism>
bool run_masked_cases() {
    return run_masked_unary_case<Parallelism>() &&
           run_masked_binary_case<Parallelism>();
}

int main() {
    if (!run_masked_cases<tsl::dataparallel::fixed<1>>()) {
        return 1;
    }
    if (!run_masked_cases<tsl::dataparallel::generic<4>>()) {
        return 2;
    }
    if (!run_masked_cases<tsl::dataparallel::generic<16>>()) {
        return 3;
    }
    return 0;
}

File: examples/rust/src/bin/masked_operator.rs
use tsl::profile;
use tsl::tsl_core::StaticSimdVector;

struct LessThan;

impl<V> profile::algo::BinaryPredicateKernel<V> for LessThan
where
    V: StaticSimdVector<BaseType = i32> + profile::detail::primitives::Less_thanImpl,
{
    fn test(&mut self, left: V::RegisterType, right: V::RegisterType) -> V::MaskType {
        profile::less_than::<V>(left, right)
    }
}

struct SquareOrOriginal;

impl<V> profile::algo::MaskedUnaryKernel<V> for SquareOrOriginal
where
    V: StaticSimdVector<BaseType = i32>
        + profile::detail::primitives::SelectImpl
        + profile::detail::primitives::MulImpl,
    V::RegisterType: Copy,
{
    fn apply(&mut self, active: V::MaskType, value: V::RegisterType) -> V::RegisterType {
        let squared = profile::mul::<V>(value, value);
        profile::select::<V>(active, squared, value)
    }
}

struct AddOrLeft;

impl<V> profile::algo::MaskedBinaryKernel<V> for AddOrLeft
where
    V: StaticSimdVector<BaseType = i32>
        + profile::detail::primitives::AddImpl
        + profile::detail::primitives::SelectImpl,
    V::RegisterType: Copy,
{
    fn apply(
        &mut self,
        active: V::MaskType,
        left: V::RegisterType,
        right: V::RegisterType,
    ) -> V::RegisterType {
        let sum = profile::add::<V>(left, right);
        profile::select::<V>(active, sum, left)
    }
}

fn fill_inputs(left: &mut [i32], right: &mut [i32]) {
    for (i, (left_value, right_value)) in left.iter_mut().zip(right.iter_mut()).enumerate() {
        *left_value = (i % 29) as i32 - 14;
        *right_value = ((i * 11) % 47) as i32 - 23;
    }
}

fn main() {
    let mut left = vec![0i32; 1000];
    let mut right = vec![0i32; 1000];
    fill_inputs(&mut left, &mut right);

    macro_rules! run_policy {
        ($policy:expr) => {{
            let policy = $policy;
            let mask_count = profile::algo::integral_mask_chunk_count::<_, i32>(policy, left.len());
            let mut masks = vec![0u64; mask_count];
            let mut less_than = LessThan;
            let produced =
                profile::algo::predicate_binary(policy, &mut less_than, &left, &right, &mut masks);
            assert_eq!(produced, masks.len());

            let unary_sentinel = -777777;
            let mut unary_output = vec![unary_sentinel; left.len()];
            let mut square = SquareOrOriginal;
            profile::algo::transform_masked_unary(
                policy,
                &mut square,
                &left,
                &masks,
                &mut unary_output,
            );

            for (i, actual) in unary_output.iter().enumerate() {
                let expected = if left[i] < right[i] {
                    left[i] * left[i]
                } else {
                    left[i]
                };
                assert_eq!(*actual, expected);
                assert_ne!(*actual, unary_sentinel);
            }

            let binary_sentinel = -888888;
            let mut binary_output = vec![binary_sentinel; left.len()];
            let mut add = AddOrLeft;
            profile::algo::transform_masked_binary(
                policy,
                &mut add,
                &left,
                &right,
                &masks,
                &mut binary_output,
            );

            for (i, actual) in binary_output.iter().enumerate() {
                let expected = if left[i] < right[i] {
                    left[i] + right[i]
                } else {
                    left[i]
                };
                assert_eq!(*actual, expected);
                assert_ne!(*actual, binary_sentinel);
            }
        }};
    }

    run_policy!(tsl::dataparallel::fixed::<1>());
    run_policy!(tsl::dataparallel::generic::<4>());
    run_policy!(tsl::dataparallel::generic::<16>());
}
